src/import/codegen.ts

```diff
@@ -81,6 +92,11 @@ export function generateConstruct(typegen: TypeGenerator, def: ApiObjectDefiniti
       copy.required = copy.required || [];
       copy.required = copy.required.filter(x => x !== 'apiVersion' && x !== 'kind' && x !== 'status');
 
+      if (def.custom) {
+        // add "metadata" field for all CRDs, overriding any existing typings
+        copy.properties.metadata = { $ref: '#/definitions/ApiObjectMetadata' };
```

Comment:
This feels a bit blunt.. Isn't it possible for a CRD to define some extra fields in their `metadata` section? Shouldn't we merge?

Reply:
From my research, defining extra fields is not possible, no. The kubernetes docs says

> if metadata is specified, then only restrictions on metadata.name and metadata.generateName are allowed.

See also this related PR: https://github.com/kubernetes/kubernetes/pull/77653

It's possible that a CRD further constraints one of these existing fields (e.g. imposing that metadata.name must start with the prefix "foo"), but I think this is exceedingly rare and the fields are still behaviorally the same. The only thing that we might lose are the descriptions of existing metadata fields that some CRD owners include in their specs - but a lot of these are redundant as is.

Reply:
Thanks for the details. Worth adding a note in the code that this is the behavior